src/Tasks/Microsoft.NET.Build.Tasks/CompileReadyToRunImages.cs

```diff
@@ -0,0 +1,376 @@
+                    MetadataReader mdReader = pereader.GetMetadataReader();
+                    foreach (var assemblyRefHandle in mdReader.AssemblyReferences)
+                    {
+                        AssemblyReference assemblyRef = mdReader.GetAssemblyReference(assemblyRefHandle);
+                        if ((assemblyRef.Flags & AssemblyFlags.WindowsRuntime) == AssemblyFlags.WindowsRuntime)
+                            return false;
+                    }
+
+                    // Skip reference assemblies
+                    foreach (var attributeHandle in mdReader.GetAssemblyDefinition().GetCustomAttributes())
+                    {
+                        EntityHandle attributeCtor = mdReader.GetCustomAttribute(attributeHandle).Constructor;
+
+                        string attributeTypeName = null, attributeTypeNamespace = null;
+
+                        if (attributeCtor.Kind == HandleKind.MemberReference)
+                        {
+                            EntityHandle attributeMemberParent = mdReader.GetMemberReference((MemberReferenceHandle)attributeCtor).Parent;
+                            if (attributeMemberParent.Kind == HandleKind.TypeReference)
+                            {
+                                TypeReference attributeTypeRef = mdReader.GetTypeReference((TypeReferenceHandle)attributeMemberParent);
+                                attributeTypeName = mdReader.GetString(attributeTypeRef.Name);
+                                attributeTypeNamespace = mdReader.GetString(attributeTypeRef.Namespace);
+                            }
+                        }
+                        else if (attributeCtor.Kind == HandleKind.MethodDefinition)
+                        {
+                            var attributeTypeDefHandle = (TypeDefinitionHandle)mdReader.GetMethodDefinition((MethodDefinitionHandle)attributeCtor).GetDeclaringType();
+                            TypeDefinition attributeTypeDef = mdReader.GetTypeDefinition(attributeTypeDefHandle);
+                            attributeTypeName = mdReader.GetString(attributeTypeDef.Name);
+                            attributeTypeNamespace = mdReader.GetString(attributeTypeDef.Namespace);
+                        }
+
+                        if (attributeTypeName == "ReferenceAssemblyAttribute" && attributeTypeNamespace == "System.Runtime.CompilerServices")
+                            return false;
+                    }
+                }
+            }
+            catch
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private bool CreateR2RImage(ITaskItem file, out string outputR2RImage)
+        {
+            outputR2RImage = Path.Combine(OutputPath, file.GetMetadata(MetadataKeys.RelativePath));
+
+            string dirName = Path.GetDirectoryName(outputR2RImage);
+            if (!Directory.Exists(dirName))
+                Directory.CreateDirectory(dirName);
```

Comment:
Directory.CreateDirectory succeeds if directory already exists, the guard is unnecessary

Reply:
Interesting... I thought it would throw an exception. Thanks for letting me know